src/openrct2-ui/drawing/engines/HardwareDisplayDrawingEngine.cpp

```diff
@@ -365,8 +365,10 @@ class HardwareDisplayDrawingEngine final : public X8DrawingEngine
 
     void RenderDirtyVisuals()
     {
-        float scaleX = Config::Get().general.WindowScale;
-        float scaleY = Config::Get().general.WindowScale;
+        int viewX, viewY;
+        SDL_GetRendererOutputSize(_sdlRenderer, &viewX, &viewY);
+        float ratioX = static_cast<float>(viewX) / static_cast<float>(_dirtyGrid.BlockColumns * _dirtyGrid.BlockWidth);
+        float ratioY = static_cast<float>(viewY) / static_cast<float>(_dirtyGrid.BlockRows * _dirtyGrid.BlockHeight);
```

Comment:
Do you get the desired outcome if you explicitly use std::round/floor/ceil?

Reply:
Using std::round, std::floor, or std::ceil causes the dirty block to become more misaligned (for example, the frame rate block), as shown in the screenshot below:
![Screenshot 2025-01-25 at 10 34 35](https://github.com/user-attachments/assets/22b54b27-2a61-4f26-b994-70ed1acc8f8d)

After testing various approaches, I found that the current implementation provides results most similar to the original behavior. Here’s a comparison

|  Current  |  Original  |
| -------- | ------- |
| <img src="https://github.com/user-attachments/assets/ab3b148d-e41d-423c-8935-db78b37a3b35" alt="New Commit" width="300"/>  | <img src="https://github.com/user-attachments/assets/75fe62a7-b569-4e07-bb0b-5e331654fabb" alt="Original" width="300"/> |

Reply:
Ahh, I think the rounding needs to take place where currently an int32 cast is used, i.e. lines 383–386. The ratio variables themselves should not be rounded.

Reply:
Ahh, I see. Let me try!

Reply:
It looks better, but there’s still some slight misalignment. Additionally, I noticed it doesn’t display well on non-Retina screens. I’ll try another solution to address this.